tests/integration/test_federation.py

```diff
@@ -0,0 +1,215 @@
+from copy import deepcopy
+import pytest
+import random
+
+import bigchaindb
+from bigchaindb.core import Bigchain
+from contextlib import contextmanager
+from bigchaindb.common.crypto import generate_key_pair
+from tests.pipelines.stepping import create_stepper
+
+
+################################################################################
+# Test setup code
+
+
+@contextmanager
+def federation(n):
+    """
+    Return a list of Bigchain objects and pipeline steppers to represent
+    a BigchainDB federation
+    """
+    keys = [generate_key_pair() for _ in range(n)]
+    config_orig = bigchaindb.config
+
+    @contextmanager
+    def make_nodes(i):
+        nonlocal keys
+        if i == 0:
```

Comment:
It's a recursive function, it is called with n where n is the number of nodes you want to simulate. When `i` is 0, there are no remaining nodes to be created so it returns.